rs/ledger_suite/icrc1/ledger/src/lib.rs

```diff
@@ -704,25 +707,45 @@ impl<Tokens: TokensType> Ledger<Tokens> {
     /// The canister code must call set_certified_data with the value this function returns after
     /// each successful modification of the ledger.
     pub fn root_hash(&self) -> [u8; 32] {
-        self.construct_hash_tree().digest().0
+        self.construct_hash_tree().digest()
     }
 
-    pub fn construct_hash_tree(&self) -> MixedHashTree {
+    pub fn construct_hash_tree(&self) -> HashTree {
         match self.blockchain().last_hash {
             Some(hash) => {
                 let last_block_index = self.blockchain().chain_length().checked_sub(1).unwrap();
-                MixedHashTree::Fork(Box::new((
-                    MixedHashTree::Labeled(
-                        Label::from("last_block_index"),
-                        Box::new(MixedHashTree::Leaf(last_block_index.to_be_bytes().to_vec())),
-                    ),
-                    MixedHashTree::Labeled(
-                        Label::from("tip_hash"),
-                        Box::new(MixedHashTree::Leaf(hash.as_slice().to_vec())),
-                    ),
-                )))
+                #[allow(unused_mut)]
+                #[allow(unused_assignments)]
+                let mut last_block_hash_label = Label::from("tip_hash");
+                let last_block_index_label = Label::from("last_block_index");
+
+                #[allow(unused_mut)]
+                #[allow(unused_assignments)]
+                let mut last_block_index_encoded = last_block_index.to_be_bytes().to_vec();
+                #[cfg(feature = "icrc3-compatible-data-certificate")]
```

Comment:
I wonder if this would be easier to read if the two implementations (with and without the feature flag) would be in their own code blocks, rather than having them interleaved. Then the `#[allow(...)]` annotations could also be dropped, and the potentially unneeded `last_block_index.to_be_bytes().to_vec()` in case the feature flag is set (since then the value is computed again and the old value overwritten). It would then also be simpler to deprecate the old behavior by simply deleting the code block with the legacy code. There may be some more duplication, but it may still be a better tradeoff.